clang/utils/TableGen/ClangProgModelBuiltinEmitter.cpp

```diff
@@ -848,6 +922,216 @@ void BuiltinNameEmitter::EmitQualTypeFinder() {
+  IsInteger = T->getValueAsBit("IsInteger");
+  IsSigned = T->getValueAsBit("IsSigned");
+  IsFloat = T->getValueAsBit("IsFloat");
+  ElementSize = T->getValueAsInt("ElementSize");
+  VecWidth = T->getValueAsInt("VecWidth");
+  IsConst = T->getValueAsBit("IsConst");
+  IsVolatile = T->getValueAsBit("IsVolatile");
+  IsPointer = T->getValueAsBit("IsPointer");
+  IsBuiltin =
+      !T->isSubClassOf("FundamentalType") && !T->isSubClassOf("VectorType");
+
+  AddrSpace = StringSwitch<const char *>(T->getValueAsString("AddrSpace"))
+                  .Case("clang::LangAS::Default", "")
+                  .Case("clang::LangAS::opencl_private", "__private")
+                  .Case("clang::LangAS::opencl_global", "__global")
+                  .Case("clang::LangAS::opencl_constant", "__constant")
+                  .Case("clang::LangAS::opencl_local", "__local")
+                  .Default("__generic");
+}
+
+std::string JSONBuiltinInterfaceEmitter::TypeDesc::GetBaseTypeAsStr() const {
+  if (Name == "void") {
+    return Name.str();
+  }
+
+  llvm::SmallString<32> Buffer;
+  {
+    llvm::raw_svector_ostream TypeStr(Buffer);
+    TypeStr << TypePrefix;
+    if (IsBuiltin) {
+      TypeStr << "_" << Name;
+    } else {
+      if (VecWidth != 1)
+        TypeStr << "_vec" << VecWidth;
+      if (IsInteger) {
+        if (ElementSize == 1)
+          TypeStr << "_bool";
+        else
+          TypeStr << (IsSigned ? "_int" : "_uint") << ElementSize;
+      } else {
+        if (IsFloat)
+          TypeStr << "_fp" << ElementSize;
+        else
+          llvm_unreachable("Unknown type");
+      }
+      TypeStr << "_t";
+    }
+  }
+
+  return static_cast<std::string>(Buffer);
+}
+
+void JSONBuiltinInterfaceEmitter::Emit() {
+
+  // Parse the Records to populate the internal lists.
+  GetOverloads();
+  GroupBySignature();
+  ExpandTypes();
+
+  EmitBuiltins();
+}
+
+// Generate all possible types represented by the input record.
+void JSONBuiltinInterfaceEmitter::ExpandType(const Record *Ty) {
+  auto ExpansionIt = ExpandedTypes.find(Ty);
+
+  if (ExpansionIt != ExpandedTypes.end())
+    return;
+  llvm::SmallVectorImpl<std::string> &Expansion = ExpandedTypes[Ty];
+
+  JSONBuiltinInterfaceEmitter::TypeDesc TypeDesc(Ty);
+
+  bool IsCompound = Ty->isSubClassOf("CompoundType");
+  if (IsCompound) {
+    auto *EltRecord = Ty->getValueAsDef("ElementType");
+    ExpandType(EltRecord);
+    Expansion = ExpandedTypes[EltRecord];
+    for (auto &TypeStr : Expansion) {
+      if (TypeDesc.IsPointer) {
+        if (TypeDesc.IsConst)
+          TypeStr += " const";
+        if (TypeDesc.IsVolatile)
+          TypeStr += " volatile";
+        TypeStr += " ";
+        TypeStr += TypeDesc.AddrSpace;
+        TypeStr += " *";
+      }
+    }
+
+    return;
+  }
+
+  if (Ty->isSubClassOf("GenericType")) {
+    assert(!Expansion.size() && "Types already expanded");
+    auto NVec = Ty->getValueAsDef("VectorList")->getValueAsListOfInts("List");
```

Comment:
Throughout the patch, you're using 'auto' in cases that the coding standard prohibits.

Reply:
fixed